Paulmck.2018.02.08a

paulmckrcu opened the pull request:
This pull request provides the following:

1. A name change to allow GCC to build successfully.
2. PowerPC assembly code which passes the self-tests.

I am new to the CPU() macro, and was unable to intuit how to specify PowerPC using that macro, so I used one of GCC's predefined macros.  Please let me know if something else would work better, but if it involves the CPU macro, I will need help.  The search engines came up empty, possibly due to lack of imagination on my part.

                                                                 Thanx, Paul

Commit:
Make gcc happy by avoiding "dependency" name collision

paulmck committed:
Add PPC code

Signed-off-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>

jfbastien commented on consume_dependency_impl.h:
```diff
@@ -38,6 +38,8 @@ inline dependency::dependency_type __create_dependency(T value) {
 #elif CPU(X86) || CPU(X86_64)
     dep = bit_cast<uint64_t>(value) ^ bit_cast<uint64_t>(value); // Any zero will do for x86
     std::atomic_signal_fence(std::memory_order_acquire);
+#elif defined(__PPC64__)
```
The CPU macro is in helpers.h. You want to add:

```
#if defined(__PPC64__)
#define WTF_CPU_PPC64
#endif
```

And here you just use

```
#elif CPU(PPC64)
```

jfbastien commented on consume.h:
```diff
@@ -119,7 +119,7 @@ class dependent_ptr {
 
     // A pure dependency from the dependent_ptr.
-    dependency dependency() const;
+    dependency Dependency() const;
```
I think this is a GCC bug? Can you rename to `dep` for now instead.

jfbastien commented on consume_dependent_ptr_impl.h:
```diff
@@ -54,6 +54,6 @@ template<typename T> inline dependent_ptr<T*> dependent_ptr<T>::operator&() cons
 template<typename T> inline T* dependent_ptr<T>::value() const { return ptr; }
 
-template<typename T> inline dependency dependent_ptr<T>::dependency() const { using shadowed = class dependency; return shadowed(ptr); }
+template<typename T> inline dependency dependent_ptr<T>::Dependency() const { using shadowed = class dependency; return shadowed(ptr); }
```
Ditto here, `dep`, and you can then remove the `shadowed` thing, just `{ return dep(ptr); }`

paulmckrcu replied:
I added a pair of commits, one to use the CPU() macro and the other to use "dep" instead of "Dependency", and also simplifying the inline function's "return" statement.

paulmck committed:
Use the CPU() macro for PowerPC

Signed-off-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>

paulmck committed:
Use "dep" rather than "Dependency"

This also allows simplifying the body of the inline function.

Suggested-by: JF Bastien <cxx@jfbastien.com
Signed-off-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>
